src/advanced_unsafety/uninitialized.md

```diff
@@ -0,0 +1,173 @@
+
+### Not exactly uninitialized: Moved-from values
+
+The following code is UB:
+
+```rust
+# use std::ptr;
+let x = String::new(); // String is not Copy
+let mut v = vec![];
+let ptr = &x as *const String;
+
+v.push(x); // move x into the vector
+
+unsafe {
+    // dangling pointer reads from moved-from memory
+    let ghost = ptr::read(ptr);
+}
+```
+
+Any type of move will do this, even when you "move" the value into a different variable with stuff like `let y = x;`.
+
+This isn't _quite_ uninitialized: it's just that using after a move is straight up UB in Rust. In particular, unlike most pointers to uninitialized values, this dangling pointer is unsound to *write* to as well.
+
+Working with dangling pointers can often lead to similar problems as working with uninitialized values.
+
+Note that Rust does let you "partially move" out of fields of a struct, in such a case the whole struct is now no longer a valid value for its type, but you are still allowed to "use" the struct to look at other fields, and the value as a whole is no longer usable. When doing such things, make sure there are no pointers that still think the struct is whole and valid.
+
+#### Caveat: `ptr::drop_in_place()`, `ManuallyDrop::drop()`, and `ptr::read()`
+
+[`ptr::drop_in_place()`] and [`ManuallyDrop::drop()`] are interesting: they both call the destructor[^2] on a value (or a pointed-to value in the case of `drop_in_place`). From the perspective of safety they are identical; they are just different APIs for dealing with manually calling destructors.
+
+[`ManuallyDrop::drop()`] makes the following claim:
+
+> Other than changes made by the destructor itself, the memory is left unchanged, and so as far as the compiler is concerned still holds a bit-pattern which is valid for the type T.
+
+In other words, Rust does _not_ consider these operations to do the same invalidation as a regular "move from" operation, even though they may have a similar feel. They do not create dangling pointers, and they do not themselves overwrite the memory with an uninitialized value.
+
+There is an [open issue][ugc-394] about whether `Drop::drop()` is itself allowed to produce uninitialized or invalid memory, so it may not be possible to rely on this in a generic context.
+
+[`ptr::read()`] similarly claims that it leaves the source memory untouched, which means that it is still a valid value. Of course, [`ptr::read()`] on a pointer pointing to uninitialized memory will still create an uninitialized value.
+
+
+For all of these APIs, actually _using_ the dropped or read-from memory may still be fraught depending on the invariants of the value; it's quite easy to cause a double-free by materializing an owned value from the original data after it has already been read-from or dropped.
+
+However, they do not produce uninitialized memory.
```

Comment:
The reason that `ManuallyDrop` makes this promise is so that calling `ManuallyDrop::drop` on the value in an `Option<ManuallyDrop<T>>` doesn't suddenly turn the option into a `None` by zeroing the pointer that the option is using for its discriminant.

I would say "always treat these as leaving the memory in an uninitialized state" and that you should never rely on them doing any more than that, other than the guarantee that niche-optimizations still work for `ManuallyDrop`.

An exception is Copy types where both are guaranteed no-ops.

Reply:
I agree that that is the right advice for unsafe code writers, but it's not an accurate description of what is happening. I'll try to expand the section a bit here.